Continuation trapping semantics

dhil opened the pull request:
This patch fixes a problem with traps on continuations, which would otherwise allow a Wasm program to continue running after invoking a trapping instruction. Currently, a fresh trap handler is installed per continuation stack, meaning that the effects of a trap is delimited by the stack segment on which the trap occurred -- whereas it really ought to be delimited by the top-level of the program (i.e. the part just before host/engine frames).

dhil committed:
Continuation trapping semantics

This patch fixes a problem with traps on continuations, which would
otherwise allow a Wasm program to continue running after invoking a
trapping instruction. Currently, a fresh trap handler is installed per
continuation stack, meaning that the effects of a trap is delimited by
the stack segment on which the trap occurred -- whereas it really
ought to be delimited by the top-level of the program (i.e. the part
just before host/engine frames).

posborne commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -866,8 +867,28 @@ impl CallThreadState {
+                    match *stack_chain {
+                        VMStackChain::Continuation(_) => true,
+                        _ => false,
+                    }
```
Should be able to be simplified slightly to just `matches!(*stack_chain, VMStackChain::Continuation(_))`

alexcrichton commented:
I'm not familiar enough with the stack-switching code currently to review this myself. For example I don't know if this is accidentally skipping over native frames at the base of other continuations. Given that I'm going to defer to @fitzgen and @posborne as they're more familiar with the details

dhil commented:
> For example I don't know if this is accidentally skipping over native frames at the base of other continuations. 

Excellent point. I think it may skip over intermediate `invoke_wasm_and_catch_traps` frames, suggesting that a "bubbling" semantics of trapping up through continuation stacks may be the right thing to do.

fitzgen reviewed (commented):
Thanks for fixing this bug!

Can we add a test that spawns an N deep stack chain with M frames where every other frame is a host frame, and the last frame (whether host or Wasm) triggers a trap? Then we can run that test exhaustively for small N and M.

Something like

```wat
(module
  ;; The imported host function.
  (import "host" "func" (func $host_func (param i32 i32)))

  ;; A global that is incremented after calling the host
  ;; function, which should trap, and therefore the
  ;; increment should never happen.
  (global $g (export "g") (mut i32) (i32.const 0))

  (func (export "run") (param $frames-per-stack i32) (param $fuel i32)
    ;; Trap on out-of-fuel for frames.
    if (i32.eqz (local.get $fuel))
      unreachable
    end

    ;; Decrement frame fuel.
    (local.set $fuel (i32.sub (local.get $fuel) (i32.const 1)))

    if (i32.eqz (i32.rem (local.get $fuel) (local.get $frames-per-stack))))
      ;; TODO: Spawn a new stack, starting either with `run`
      ;; or our host function (based on another param or a
      ;; global or something), and switch to it...
    else
      ;; Call the host function to continue our mutual recursion.
      (call $host_func (local.get $frames-per-stack) (local.get $frame-fuel))
    end

    ;; Increment the global. Should never execute, dynamically.
    (global.set $g (i32.add (global.get $g) (i32.const 1))
  )
)
```

```rust
let host_func = Func::wrap(
    &mut store,
    |mut caller: Caller<'_, ()>, frames_per_stack: u32, fuel: u32| -> Result<()> {
        ;; Trap on out-of-fuel for frames.
        if fuel == 0 {
            bail!("out of frame fuel");
        }

        ;; Mutual recursion back into the Wasm function.
        let run = instance.get_typed_func::<(u32, u32)>(&mut caller).unwrap();
        run.call(&mut caller, (frames_per_stack, fuel - 1))?;

        ;; Increment the global. Should never execute, dynamically.
        let g = instance.get_global("g").unwrap();
        let g_val = g.get(&mut caller).unwrap_i32();
        g.set(&mut caller, Val::I32(g_val + 1))?;

        Ok(())
    },
);

// ...

for frames_per_stack in 1..4 {
    for fuel in 0..frames_per_stack * 3 {
        let mut store = Store::new(&engine, ());
        let instance = Instance::new(...)?;

        let run = instance.get_typed_func::<(u32, u32)>(&mut store)?;
        run.call(&mut store, (frames_per_stack, fuel))?;

        let g = instance.get_global(&mut store, "g").unwrap();
        assert_eq!(g.unwrap_i32(), 0);
    }
}
```

This would give me a lot more confidence that we are properly handling traps across stacks, regardless of the stack chain, host functions, and what kind of frame is youngest or oldest.

(And when we add embedder API support for spawning stacks, we should also extend the host function in this new test to use that support)

dhil committed:
Implement Paul's feedback

dhil committed:
Host frames and continuation frames interpersed trap tests

dhil committed:
Propagate traps through continuation parent stacks

This patch implements a bubble-semantics for propagating traps through
nested continuation stacks. Every trap in a continuation is now caught
by its parent stack, which re-raises the trap to tis parent until the
trap is ultimately caught by the nearest delimiting host/engine
frame. In order for this bubbling to correct, we now need to record
the trap-handler state of each continuation stack such that it can be
restored accordingly before propagating the trap.

This design allows continuation bootstrap frames to return normally
and run Rust destructors, and thus is more robust than the previously
proposed design which would perform one jump to the nearest delimiting
host/engine frame.

dhil commented:
@fitzgen PTAL when its convenient -- and apologies for the latency! I have added two new commits:
* https://github.com/bytecodealliance/wasmtime/pull/11717/commits/9bae46738f14a78aec65a3adb724cf873bbbf34e: The test you asked for.
* https://github.com/bytecodealliance/wasmtime/pull/11717/commits/00715fd28d884e4aead5b0613281c409eebd49cd: A more robust way to handle trapping in nested continuations by way of "bubbling" traps through each continuation rather than jumping straight to the nearest delimiting host/engine frame. It is more robust in the sense that it will run any Rust destructor that may be in the initial continuation frame (currently there are none, I believe, but I suppose this may change in the future).

If we go with this "bubbling" approach, then I will drop the first two commits of this PR.

fitzgen commented on crates/cranelift/src/func_environ/stack_switching/instructions.rs:
```diff
@@ -809,11 +822,35 @@ pub(crate) mod stack_switching_helpers {
+                our_memflags,
+                stack_limits_ptr,
+                i32::from(last_wasm_entry_sp_offset),
+            );
```
Can you define `AliasRegions::last_wasm_entry_sp` and use it via `env.alias_regions.last_wasm_entry_sp` here? That will ensure that we are using the correct alias region for this load.

I know it is currently a little tedious, but I plan on tackling https://github.com/bytecodealliance/wasmtime/issues/13707 soon, which should remove all the boilerplate.

dhil replied:
Yes! That's an oversight from me. I have added them in https://github.com/bytecodealliance/wasmtime/pull/11717/commits/2249a9417b89c1777e9b1c33ac79714e43f623f8.

fitzgen commented on crates/cranelift/src/func_environ/stack_switching/instructions.rs:
```diff
@@ -809,11 +822,35 @@ pub(crate) mod stack_switching_helpers {
+                our_memflags,
+                stack_limits_ptr,
+                i32::from(last_wasm_entry_trap_handler_offset),
+            );
```
And similar here

fitzgen reviewed (approved):
Thanks! LGTM with nitpicks below addressed

dhil committed:
Alias region loads for continuation VMStackLimits fields

fitzgen reviewed (approved):
Thanks!

fitzgen commented:
@dhil I think you need to mark the new tests as `#[cfg_attr(miri, ignore)]`, since they run Wasm and MIRI can't handle that, except in very limited cases.

https://github.com/bytecodealliance/wasmtime/actions/runs/30111260793/job/89541246907#step:6:654

dhil committed:
asan and miri ignore continuation trap tests

prtest:full

dhil commented:
> @dhil I think you need to mark the new tests as `#[cfg_attr(miri, ignore)]`, since they run Wasm and MIRI can't handle that, except in very limited cases.
> 
> https://github.com/bytecodealliance/wasmtime/actions/runs/30111260793/job/89541246907#step:6:654

I've added the directive to the continuation trap tests in commit https://github.com/bytecodealliance/wasmtime/pull/11717/commits/ec6cf0b17afcaa6adc3f9fd50c596c5dde6725f6.

dhil commented:
@fitzgen looks like the merge queue failed on `Test macOS arm64 (wasmtime)` -- I am not sure how it is related to this PR, are there more conditional compilation directives that I need to add?

fitzgen commented:
I re-enqueued to see if that was some flaky test, but I'm not aware of any flaky segfaults (that would be pretty concerning) so I suspect that the segfault is related to this PR

dhil commented:
@fitzgen The continuation trapping tests were failing with ASAN build. I have added `ignore(asan)` directives to the two tests too. We need to instrument the stack switching runtime with asan markers in a similar way to the async fiber runtime. Though, I prefer to defer this to a subsequent patch.

There is a `wasi-nn` test that keeps failing, but I see it failed on `main` during the night too https://github.com/bytecodealliance/wasmtime/actions/runs/30326568601/job/90173165029 -- I suppose this failure is unrelated to this PR.

alexcrichton commented:
Yeah wasi-nn is no longer gated on right now while it's flaky
